service/history/workflow/cache/cache.go

```diff
@@ -164,10 +166,27 @@ func newCache(
 		TTL: ttl,
 		Pin: true,
 		OnPut: func(val any) {
-			// TODO: will be used in follow-up PR
+			item := val.(*cacheItem)
+			wfKey := item.wfContext.GetWorkflowKey()
+			err := item.finalizer.Register(wfKey.String(), func(ctx context.Context) error {
+				if err := item.wfContext.Lock(ctx, locks.PriorityHigh); err != nil {
+					return err
+				}
+				defer item.wfContext.Unlock()
+				item.wfContext.Clear()
+				return nil
+			})
+			if err != nil {
+				logger.Warn("finalizer failed to register callback", tag.Error(err))
```

Comment:
There is not much we can do if any of these operations fail, other than logging it.